client/src/util/dataframe/types.ts

```diff
@@ -0,0 +1,145 @@
+export type CategoricalHistogramBy = Map<DataframeValue, CategoricalHistogram>;
+
+export type DataframeValue = number | string | boolean;
+
+export type DataframeValueArray = DataframeValue[] | TypedArray;
+
+export type DataframeColumnGetter = (
+  label: LabelType
+) => DataframeValue | undefined;
+
+/**
+ * Interface representing a Dataframe column.  Eg, returned by
+ * Dataframe.col().
+ */
+export interface DataframeColumn extends DataframeColumnGetter {
+  /**
+   * __id is unique per Dataframe and DataframeColumn, and is used as a memoization key.
+   */
+  readonly __id: string;
+
+  /**
+   * Boolean indicating if the underlying data supports continuous operations, eg,
+   * summarizeContinuous.
+   */
+  isContinuous: boolean;
+
+  /**
+   * Return underlying column data as an array-like object.
+   */
+  asArray: () => DataframeValueArray;
+
+  /**
+   * Continuous data summary.  Will throw if !isContinuous.
+   */
+  summarizeContinuous: () => ContinuousColumnSummary;
+
+  /**
+   * Categorical data summary.
+   */
+  summarizeCategorical: () => CategoricalColumnSummary;
+
+  /**
+   * Continuous bin/histogram.  Will throw if !isContinuous.
+   * @param bins - array of bin boundary fractions, in range [0., 1.]
+   * @param domain - data domain [min, max]
+   */
+  histogramContinuous: (
+    bins: number,
+    domain: [number, number]
+  ) => ContinuousHistogram;
+
+  /**
+   * Continuous bin/histogram, grouped by another categorical column.  Will throw if !isContinuous.
+   * @param bins - array of bin boundary fractions, in range [0., 1.]
+   * @param domain - data domain [min, max]
+   * @param by - group by categorical column
+   */
+  histogramContinuousBy: (
+    bins: number,
+    domain: [number, number],
+    by: DataframeColumn
+  ) => ContinuousHistogramBy;
+
+  /**
+   * Categorical bin/histogram.
+   */
+  histogramCategorical: () => CategoricalHistogram;
+
+  /**
+   * Categorical bin/histogram grouped by another column.
+   * @param by - group by categorical column
+   */
+  histogramCategoricalBy: (by: DataframeColumn) => CategoricalHistogramBy;
+
+  /**
+   * Return true if the column contains the row label.
+   */
+  has: (rlabel: LabelType) => boolean;
+
+  /**
+   * Return true if the column contains the row offset. Identical to
+   * (offset >= 0 && offset < dataframe.length)
+   */
+  ihas: (offset: OffsetType) => boolean;
+
+  /**
+   * Return index of the value, as a _label_. Returns undefined if
+   * not present.  *NOTE*: unlike Array.indexOf, does not return an
+   * offset.
+   */
+  indexOf: (value: DataframeValue) => LabelType | undefined;
+
+  /**
+   * Return the value at the given offset, or undefined if not present.
+   */
+  iget: (offset: OffsetType) => DataframeValue | undefined;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any --- a legitimate use of any.
+export type AnyFunction<T = unknown> = (...args: any[]) => T;
```

Comment:
just curious if `unknown[]` would work here instead of `any[]`?

https://blog.logrocket.com/when-to-use-never-and-unknown-in-typescript-5e4d6c5799ad/

Reply:
No, it won't work as you can't call the memoized function without a type error.  See, for example, how TS defines the signatures to Function.call, Function.bind, etc.

What _would_ be better is to improve the type pass through for the memoize() function, which is the only user of this type.  I'll fix that for both memoize() and callOnceLazy().

Eg,
```
export function memoize<
  T extends (...args: any[]) => any = (...args: any[]) => any
>(
  fn: T,
  hashFn: (...args: Parameters<T>) => string,
  maxResultsCached = -1
): (...args: Parameters<T>) => ReturnType<T> {
    ...
}
```

Side note: fixing this pointed out a small typing bug!  Yay.

Reply:
ps. if you haven't already done so, it is worth studying the type signatures for stuff like Function.bind, Function.apply, etc.

Reply:
Ahh didn't think of checking out those signatures! Thanks for the tip 👍 

Using `Parameters<T>` SGTM!